To-do 테스트 작성하기

2scent opened the pull request:
안녕하세요.  3주 차 과제 To-do 테스트 작성하기 리뷰 요청드립니다.

우선은 `Input` 컴포넌트 테스트만 작성해 봤는데요. 다른 컴포넌트도 시도는 해봤는데 `Input`  컴포넌트와 달리 컴포넌트 안에 컴포넌트가 있는 구조다 보니 테스트를 어떤 식으로 작성해야 할지 감이 잘 안 잡히네요.

예를 들어 `List` 컴포넌트의 경우, `props`로 받은 `tasks`를 모두 출력하고, 각 항목들이 `onClickDelete`를 가져야 한다는 것을 테스트해야 할 것 같은데 이런 방향성이 맞을까요?

2scent committed:
Add Input test

passwd10 commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,41 @@
+
+describe('Input', () => {
+  const value = '지금 할 일';
+  const handleChange = jest.fn();
+  const handleClick = jest.fn();
```
mock 함수를 사용하고 초기화해주지 않으면 다른 테스트 케이스에도 영향을 줄 수 있기 때문에 mock 함수를 정리해주는것이 좋아요.

beforeEach 와 mock clear 를 찾아서 적용해보세요!

passwd10 commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,41 @@
+  });
+
+  it('has label', () => {
+    expect(screen.getByText('할 일')).toBeInTheDocument();
```
`screen` 을 사용하신 이유가 있을까요?

2scent replied:
`render()`를 `beforeEach()`에서만 실행하려고 하니깐 각 요소에 접근할 방법이 마땅치 않아 screen을 사용했는데 각 test마다 `render()`를 따로 호출하는 게 나은 방법일까요?

passwd10 replied:
각각의 테스트는 독립적으로 실행되기 때문에 render 가 반복되어 호출될 수 밖에 없어요! 
매번 호출하는게 번거로우면 상위에 컴포넌트를 render하는 함수로 분리해서 표현할 수 있을 것 같아요.

passwd10 commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,41 @@
+
+  it('shows value in props', () => {
+    expect(screen.getByDisplayValue(value)).toBeInTheDocument();
+  });
```
`describe` - `context` - `it` 패턴을 사용해서 테스트 상황을 구체화하시면 좋을 것 같아요.

예를들면 `Input이 ~~상황일 때 ~~동작할 것이다.` 같은 형식으로요

passwd10 commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,41 @@
+        onClick={handleClick}
+      />,
+    );
+  });
```
```javascript
const { container } = render(<input ... />)
```

형식으로 컴포넌트를 선언하시고, render 된 컴포넌트로 부터 반환된 `container`, `getBy~~` 등을 활용해서 테스트를 작성해보세요

passwd10 commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,41 @@
+    expect(screen.getByDisplayValue(value)).toBeInTheDocument();
+  });
+
+  it('has input with placeholder and calls onChange when input is change', () => {
```
테스트 케이스에 너무 많은 조건이 포함되어 있어요. 하나의 상황만 가정하고 간결하게 테스트를 작성해보세요 :)

passwd10 commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,41 @@
+    expect(screen.getByText('할 일')).toBeInTheDocument();
+  });
+
+  it('shows value in props', () => {
```
사용자의 관점에서 테스트케이스를 작성해보세요. 다음과같은 상황에선 `renders text` 로 표현하면 좋겠죠?

passwd10 replied:
`renders text` 로 표현하면 컴포넌트의 동작, 역할을 더 확실히 알 수 있을 것 같아요

passwd10 reviewed (commented):
테스트코드 작성이 처음이시면 무엇을 어떻게 테스트해야하는지 헷갈리실거에요 ㅎㅎ 
강의에 나온 패턴을 반복해서 따라해보시고 그대로 과제에 적용해보세요!

> 우선은 Input 컴포넌트 테스트만 작성해 봤는데요.
다른 컴포넌트도 시도는 해봤는데 Input 컴포넌트와 달리 컴포넌트 안에
컴포넌트가 있는 구조다 보니 테스트를 어떤 식으로 작성해야 할지 감이 잘 안 잡히네요.

각 컴포넌트에서 테스트를 통해 무엇을 확인하고싶은지 고민해보시면 좋을 것 같아요!

2scent committed:
Edit Input test

render()를 매번 호출할 때 중복을 제거하기 위해 setUp() 함수 선언
beforeEach()에서 mockClear() 호출
describe - context - it 패턴 적용

2scent committed:
Add List test

2scent commented:
피드백 주신대로 고민해봤는데 여전히 각 컴포넌트에서 무엇을 테스트해야 할지 잘 모르겠습니다.

`List` 컴포넌트의 역할은 props로 받은 `tasks`만큼 `Item` 컴포넌트를 보여주는 거라 생각하는데요.

```javascript
tasks.forEach((task) => {
  expect(container).toHaveTextContent(task.title);
});
```

이런 식으로 테스트를 작성하면 `Item` 컴포넌트의 구현을 제약하거나 의존적인 테스트가 되는 것 같은데 이게 맞는 방향일까요?

같은 이유로 `Page` 컴포넌트는 `To-do`를 보여줘야 한다 말고는 무엇을 테스트해야 할지 감이 안 잡히네요.

passwd10 commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,54 @@
+
+  beforeEach(() => {
+    handleChange.mockClear();
+    handleClick.mockClear();
```
```javascript
 jest.clearAllMocks();
```
라는 메소드를 사용하면 한번에 clear 가 가능합니다 ㅎㅎ

passwd10 commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,54 @@
+      onChange={handleChange}
+      onClick={handleClick}
+    />,
+  );
```
함수로 잘 분리하셨어요 👍

passwd10 commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,54 @@
+  });
+
+  context('when input is changed', () => {
+    it('calls onChange', () => {
```
`context` 는 상반되는 상황에서 문맥을 분기처리할 때 사용하는 표현이에요.

`listens value change event` 라고 표현해줘도 충분할 것 같아요 :)

passwd10 commented on src/List.test.jsx:
```diff
@@ -0,0 +1,43 @@
+    });
+  });
+
+  context('with tasks', () => {
```
context 를 적절하게 잘 사용하셨어요 💯

passwd10 commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,54 @@
+import Input from './Input';
+
+describe('Input', () => {
+  const value = '지금 할 일';
```
만약 value가 비어있는 경우를 테스트하고 싶은상황이 오면, 현재 전역으로 선언한 value를 사용하지 못하겠죠? 

테스트 안에서 선언해서 넣어주시면 유연하게 테스트를 작성할 수 있을 것 같아요!

passwd10 reviewed (commented):
컴포넌트를 직접 실행하지 않고도, 테스트코드만을 보고 컴포넌트의 역할과 동작을 설명할 수 있는 명세를 작성한다고 생각해보시면 도움이 될 것 같아요.

Page 컴포넌트를 예로들어볼게요.
Page 컴포넌트의 행동(역할)을 나열하면 다음과 같아요

```
- title을 렌더링한다
- input을 렌더링한다
- change event를 listen한다
- 등..
```

나열한것들이 이미 하위컴포넌트에서 테스트한것일수도 있지만 상관없어요.
Page 컴포넌트에서도 발생하고있는 일들이니까요.
각각의 컴포넌트를 독립적으로 바라보세요 ㅎㅎ
이제 나열한것들을 그대로 테스트 시나리오로 옮겨서 작성하시면 돼요
 
질문주신 List 컴포넌트의 경우에도, tasks만큼 title을 보여주는것이 List 의 역할이라면 그대로 테스트를 작성하시면 돼요! 상위컴포넌트가 하위컴포넌트를 사용하고 있으니 당연히 영향을 받을수밖에 없어요. Item 이 변경되어서 List의 테스트가 깨진다면 올바로 테스트를 작성하고 계신거에요 😄

2scent committed:
Edit Input test

불필요하게 나눈 when - it을 it으로 통합
setUp()에서 value를 유동적으로 받을 수 있도록 수정

2scent committed:
Edit List test

- tasks 렌더에 대한 테스트를 각 tasks의 title이 올바르게 렌더됐는지
테스트하는  거로 수정
- 클릭 이벤트 테스트 추가

2scent committed:
Add Page test

2scent commented:
피드백 주신 내용 참고해서 `Input`, `List` 컴포넌트 테스트 수정하고, `Page` 컴포넌트 테스트를 추가해봤는데 잘 적용한 건지 모르겠네요.. 🤔

`App` 컴포넌트에 대한 아직 테스트는 작성하지 못했고, 피드백 주신 것처럼 명세부터 작성한다는 느낌으로 생각해보고 있는데요.

```
- taskTitle을 렌더링한다.
- input이 change 되면 주면 taskTitle이 바뀐다.
- 추가 버튼을 click 하면,
  - taskTitle을 빈 문자열('')로 만든다.
  - taskTitle을 렌더링한다.
- 삭제 버튼을 click 하면,
  - 그에 맞는 taskTitle을 화면에서 지운다.
```

이런 식으로 하면 될까요? 테스트 할 내용이 많아지니깐 context에 어떤 내용이 들어가야 하는지 좀 헷갈리네요.

마지막으로 질문이 있는데요. 이번에 과제를 진행하면서 버튼을 가져오는 방법으로 `getByText`랑 `getByRole` 둘다 써봤는데 이렇게 다양한 방법이 있는데 좀 더 선호되는 방법이 있을까요?

passwd10 commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,49 @@
+  const handleChange = jest.fn();
+  const handleClick = jest.fn();
+
+  const setUp = (value = '') => render(
```
`renderInput` 으로 표현하면 함수의 역할이 더 잘 드러날 것 같아요!

passwd10 commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,49 @@
+
+    expect(handleClick).not.toBeCalled();
+    fireEvent.click(getByText('추가'));
+    expect(handleClick).toBeCalled();
```
각각 맥락이 다르기 때문에 공백으로 구분해주시면 좋을 것 같아요.

passwd10 reviewed (commented):
감을 잘 잡고 진행하고계신 것 같아요 :)
테스트내용이 많아져도 상반된 상황이 아니면 굳이 context를 사용할 필요는 없어보여요!

전 `getByText`를 사용했을 때 element가 명시적으로 더 잘 드러나는 것 같아서 `getByText`를 선호하는 편이에요.
element의 구조적인 역할에 조금 더 집중하고 싶다면 `getByRole`을 사용하셔도 될 것 같아요
다양한 `getBy~` 메소드를 사용해보시면서 실험하는 자세가 정말 훌륭합니다 👍  

추가로 App 컴포넌트에 대한 테스트도 작성해보세요!

2scent committed:
Edit List test

setUp() -> renderList()
클릭 이벤트 테스트에서 불필요한 expect 삭제

2scent committed:
Edit Input test

setUp() -> renderInput()
이벤트 테스트에서 단언을 expect(handler).toBeCalledTimes(1)로 통일

2scent committed:
Edit Page test

setUp() -> renderPage()
불필요한 toBeCalled() 제거하고, toBeCalledTimes()로 통일

nakyeonko3 commented:
과제 정말 열심히 하셨네요. 참고 좀 하겠습니다.

2scent committed:
Add App test

2scent commented:
피드백 참고해서 `App` 컴포넌트 테스트도 작성해봤는데 막상 해보니깐 역시 어렵네요. 이렇게 하는 게 맞는 건지 잘 모르겠는데 이번에도 많은 피드백 부탁드립니다.

---

그동안 주신 피드백을 봤을 때 트레이너님은 `screen`을 사용하는 것보다 `render`의 리턴값을 이용해서 쿼리하는 것을 권장하는 것 같다는 생각이 드는데요. 두 방법의 차이를 검색하다가 React Testing Library의 개발자인 Kent C. Dodds의 글을 봤는데,

> The benefit of using `screen` is you no longer need to keep the `render` call destructure up-to-date as you add/remove the queries you need. You only need to type `screen`. and let your editor's magic autocomplete take care of the rest.

[https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#not-using-screen](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#not-using-screen)

그는 필요한 쿼리를 추가/제거 할 때 `render`를 호출해 최신 상태를 유지할 필요가 없다는 이유로 `screen` 사용을 권장하더라고요. 혹시 트레이너님께서 `render`의 리턴값을 이용하는 방법을 더 권장하는 이유를 알 수 있을까요?

passwd10 commented on src/Page.test.jsx:
```diff
@@ -0,0 +1,81 @@
+  });
+
+  context('with tasks', () => {
+    let tasks;
```
let을 사용하지 않고 테스트를 작성해보세요!

2scent replied:
제 코드를 다시 보니깐 다른 곳에선 다 `const`를 사용했는데 여기서만 일관성 없게 `let`을 사용했었네요.

제가 왜 여기서 `let`을 썼는지 떠올려보니깐 저 코드를 작성할 때, "`tasks`가 테스트 도중에 바뀌면 어쩌지?"라는 생각을 했던 것 같아요. `const`로 선언한다고 해도 배열이다 보니 변환할 방법은 많으니까요.

그래서 하는 질문인데요. 지금 코드처럼 `let`으로 변수를 선언하고, `beforeEach()`에서 매번 초기화화는 방법 말고, 여러 테스트에서 사용되는 값이 변하지 않았음을 확신하는 방법으로 또 뭐가 있을까요?

```javascript
const getTasks = () => [
  { id: 1, title: '할 일 1' },
  { id: 2, title: '할 일 2' },
  { id: 3, title: '할 일 3' },
];
```

이런 함수를 선언하고, 매 테스트에서 호출하는 방법도 생각해봤는데 오히려 가독성이 떨어지는 것 같아요.

passwd10 commented on src/App.test.jsx:
```diff
@@ -0,0 +1,92 @@
+
+    it('removes task when complete button is clicked', () => {
+      const { container, getAllByText } = renderAppWithTasks();
+      const completeButton = getAllByText('완료');
```
복수형으로 표현해주세요!

passwd10 commented on src/App.test.jsx:
```diff
@@ -0,0 +1,92 @@
+      '할 일 5',
+    ];
+
+    const renderAppWithTasks = () => {
+      const result = renderApp();
+      const { getByRole, getByText } = result;
+
+      taskTitles.forEach((taskTitle) => {
+        fireEvent.change(getByRole('textbox'), { target: { value: taskTitle } });
+        fireEvent.click(getByText('추가'));
+      });
+
+      return result;
+    };
```
반복되는 부분을 함수로 잘 분리하신 것 같아요 👍

passwd10 commented on src/List.test.jsx:
```diff
@@ -0,0 +1,54 @@
+      tasks={tasks}
+      onClickDelete={handleClick}
+    />,
+  );
```
List를 괄호로 감싸면 콤마가 필요없어져요.

passwd10 commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,49 @@
+
+    fireEvent.change(getByPlaceholderText('할 일을 입력해 주세요'), { target: { value: 'a' } });
+
+    expect(handleChange).toBeCalledTimes(1);
```
테스트 시나리오대로 event listen에 대한 확인만 하는거면 `toBeCalled()` 로 표현해도 될 것 같아요

passwd10 reviewed (commented):
현재 App 컴포넌트 내부에 로직이 구현되어있기 때문에 상태나 로직에 대한 테스트를 진행하긴 어려우실거에요. 따라서 E2E 테스트와 비슷하게 테스트를 작성하시면돼요!
상태나 로직에 대한 테스트는 다음 과제에서 Redux를 통해 진행하시게 될거에요.

말씀하신대로 screen의 장점이 있긴하지만 매번 `screen.` 을 붙여줘야한다는 점이
코드의 가독성을 떨어뜨린다고 생각이되어서 개인적으로 screen 잘 사용하지 않는 편이에요.
이 부분은 정답이 있는것은 아니고 각자의 판단과 기준에 따라 달라질 수 있을 것 같아요 :)

2scent committed:
Edit Input test to use toBeCalled()

2scent committed:
Refactor Input test

renderInput()에서 Input을 표현할 때, 콤마 대신 괄호 사용

2scent committed:
Refactor List test

renderList()에서 콤마 대신 괄호 사용

2scent committed:
Refactor Page test

renderPage()에서 콤마 대신 괄호 사용

2scent committed:
Edit Page test to use toBeCalled()

2scent committed:
Edit names of some tests in App test

2scent commented:
피드백 참고해서 코드에 반영했고, `let` 사용과 관련된 부분은 답변 주시면 좀 더 생각해보려고 우선은 수정하지 않고 놔뒀습니다. 이번 리뷰도 잘 부탁드립니다 😄

passwd10 reviewed (commented):
tasks는 props 로 넘기기 위해 사용되는 임시 변수이기 때문에 테스트 도중에 바뀔일이 없지 않을까요?
만약 상황에 따라 다른 tasks 값을 사용해야한다면, 그 땐 테스트를 분리해서 작성하거나 `it` 내부에서 변수를 선언하면 될 것 같아요.
또는 `given2` 라는 라이브러리를 사용해서 상황에 맞게 lazy하게 값을 지정하는 방법도 있어요!

피드백 드린부분들 잘 적용해주셨네요! 더 드릴 피드백은 없어보여요 ㅎㅎ
앞으로 남은 과제에서도 테스트를 작성할일이 많으니 이번에 알게된 내용들 놓치지 않고 꾸준히 연습하시면 좋을 것 같아요 :)

2scent committed:
Refactor Page test to remove let

2scent commented:
이번 주 회고입니다.

[https://han-study.tistory.com/36](https://han-study.tistory.com/36)

꼼꼼하게 좋은 피드백을 주신 덕분에 테스트에 대한 감이 잡힌 것 같습니다. 감사합니다 🙏

passwd10 commented:
과제에서 많은 고민을 하신만큼 얻어가는게 많으셨을거라 생각해요
지금처럼 다양하게 시도하시고 꾸준히 의도를 물어가면서 코드숨 과제를 진행하시면 좋을 것 같아요 😄 

이번주도 고생 많으셨고 다음주도 화이팅입니다! 💪💪
